File: Erlang/swapview.erl
-module(swapview).
-export([swap_print/0, getswapfor/2]).

is_integer(S) ->
    case string:to_integer(S) of
        {_, []} ->
            true;
        _ ->
            false
    end.

filesize(Size) ->
    Units = "KMGT",
    case filesize(Size, 0) of
        {Fixsize, 0} ->
            io_lib:format("~.0fB", [Fixsize]);
        {Fixsize, Unit} when Unit > 0 ->
            io_lib:format("~.1f~ciB", [Fixsize, lists:nth(Unit, Units)])
    end.

filesize(Size, Unit) 
  when Size < 0 ->
    {Size1, Unit1} = filesize(-Size, Unit),
    {-Size1, Unit1};
filesize(Size, Unit)
  when Unit < 4, Size > 1100 ->
    filesize(Size / 1024, Unit + 1);
filesize(Size, Unit) ->
    {Size, Unit}.

swap_print() ->
    Ret = getswap(),
    io:format("~5s ~9s ~s~n", ["PID", "SWAP", "COMMAND"]),
    [io:format("~5s ~9s ~s~n", [Pid, filesize(Size), Cmd]) || {Pid, Cmd, Size} <- Ret],
    Total = lists:sum([Size || {_,_,Size} <- Ret]),
    io:format("Total: ~8s~n", [filesize(Total)]).

getswap() ->
    {ok, F} = file:list_dir_all("/proc"),
    Pids = lists:filter(fun is_integer/1, F),
    [spawn(?MODULE, getswapfor, [self(), Pid]) || Pid <- Pids],
    S = [read_result() || _ <- Pids],
    lists:keysort(3, [E || E = {_,_,Size} <- S, Size > 0]).

read_result() ->
    receive
        Data ->
            Data
    end.

read_file(File) ->
    case file:read(File, 4096) of
        {ok, Data} ->
            Data ++ read_file(File);
        eof ->
            []
    end.

getswapfor(Master, Pid) ->
    case file:open("/proc/"++Pid++"/cmdline", [read, raw, read_ahead]) of
        {ok, File} ->
            Cmd =
                try
                    string:join(string:tokens(read_file(File), [0]), " ")
                after
                    file:close(File)
                end,

            case file:open("/proc/"++Pid++"/smaps", [read, raw, read_ahead]) of
                {ok, File2} ->
                    Size =
                        try
                            getswapsize(File2, 0) * 1024
                        after
                            file:close(File2)
                        end,
                    Master ! {Pid, Cmd, Size};
                {error, _} ->
                    Master ! 0
            end;
        {error, _} ->
            Master ! 0
    end.

getswapsize(File, Acc) ->
    case file:read_line(File) of
        eof -> Acc;
        {ok, "Swap:" ++ Rest} ->
            {Size, _} = string:to_integer(string:strip(Rest, left, 32)),
            getswapsize(File, Acc+Size);
        {ok, _} ->
            getswapsize(File, Acc)
    end.

File: R/swapview.r
#!/usr/bin/env Rscript

filesize <- function (size){
  left = abs(size)
  unit = 0
  units = unlist(strsplit("KMGT",""))
  while ( left > 1100 && unit < 4){
    left = left / 1024
    unit = unit + 1
  }
  ret = ""
  if( unit == 0 ){
    sprintf("%dB", size)
  }else{
    sprintf("%.1f%siB", left, units[unit])
  }
}

getComm <- function (pid){
  fcomm = file(sprintf('/proc/%d/cmdline', pid), "rb")
  on.exit(close(fcomm))
  comm = NULL
  rawspace = charToRaw(" ")
  while(length(cbin <- readBin(fcomm, raw())) != 0){
    comm = c(comm, if(cbin[1] == raw(1)) rawspace else cbin[1])
  }
  comm = rawToChar(comm[1:length(comm) - (if(comm[length(comm)]==rawspace) 1 else 0)])
  comm
}

getSize <- function(pid){
  fsmaps = file(sprintf('/proc/%d/smaps', pid), "rb")
  on.exit(close(fsmaps))
  smaps = readLines(fsmaps)
  sum(sapply(smaps[grepl("Swap:", smaps)],
             function(a) as.numeric(rev(unlist(strsplit(a, " ")[[1]]))[2]))
      )*1024
}

getSwapFor <- function(pid){
  tryCatch(list(pid, getSize(pid), getComm(pid)),
           error=function(e) list(pid, 0, ""))
}

getSwap <- function(){
  all = sapply(sapply(list.files("/proc", pattern="^[0-9]+$"), strtoi), getSwapFor)
  all = all[, all[2,] > 0]
  all = data.frame(pid=unlist(unname(all[1,])), 
                   size=unlist(unname(all[2,])), 
                   comm=unlist(unname(all[3,])))
  if(length(all)>0){
      all = all[order(all[,2]),]
  }
  all
}

result = getSwap()
write(sprintf("%7s %9s %s", "PID", "SWAP", "COMMAND"), stdout())
t = 0
for(i in 1:nrow(result)){
  size = as.numeric(result[i,2])
  if(length(size)>0){
    write(sprintf("%7s %9s %s", result[i,1], filesize(size), result[i,3]), stdout())
    t = t + size
  }
}
write(sprintf("Total: %10s", filesize(t)), stdout())
